packages/arb-bridge-peripherals/contracts/tokenbridge/arbitrum/ArbTokenBridge.sol

```diff
@@ -213,66 +123,99 @@ contract ArbTokenBridge is ProxySetter {
+                _withdraw(l1ERC20, sender, amount);
+                return;
+            }
+        }
+        // ignores deployData if token already deployed
 
         if (callHookData.length > 0) {
-            handleCallHookData(token, amount, sender, dest, callHookData);
+            handleCallHookData(expectedAddress, amount, sender, dest, callHookData);
         } else {
-            token.bridgeMint(dest, amount, "");
+            IArbToken(expectedAddress).bridgeMint(dest, amount);
         }
 
-        emit TokenMinted(
-            l1ERC20,
-            address(token),
-            tokenType,
-            sender,
-            dest,
-            amount,
-            callHookData.length > 0
-        );
+        emit TokenMinted(l1ERC20, expectedAddress, sender, dest, amount, callHookData.length > 0);
     }
 
-    function updateTokenInfo(
-        address l1ERC20,
-        StandardTokenType tokenType,
-        bytes calldata _name,
-        bytes calldata _symbol,
-        bytes calldata _decimals
-    ) external onlyEthPair noCustomToken(l1ERC20) {
-        // no custom token as we assume custom implementation has correct info
-        require(tokenType != StandardTokenType.Custom, "Cant update info of custom token");
-        string memory name = BytesParserWithDefault.toString(_name, "");
-        string memory symbol = BytesParserWithDefault.toString(_symbol, "");
-        uint8 decimals = BytesParserWithDefault.toUint8(_decimals, 18);
-
-        IArbToken token = ensureTokenExists(l1ERC20, decimals, tokenType);
-        token.updateInfo(name, symbol, decimals);
-
-        emit TokenDataUpdated(l1ERC20, address(token), tokenType, name, symbol, decimals);
+    function deployToken(address l1ERC20, bytes memory deployData) internal returns (address) {
+        address beacon = templateERC20;
+
+        deployBeacon = beacon;
+        bytes32 salt = keccak256(abi.encodePacked(l1ERC20, beacon));
+        address createdContract = address(new ClonableBeaconProxy{ salt: salt }());
+        deployBeacon = address(0);
+
+        bool initSuccess = IArbToken(createdContract).bridgeInit(l1ERC20, deployData);
+        assert(initSuccess);
+
+        emit TokenCreated(l1ERC20, createdContract);
+        return createdContract;
     }
 
-    function customTokenRegistered(address l1Address, address l2Address) external onlyEthPair {
-        customToken[l1Address] = l2Address;
-        emit TokenCreated(l1Address, l2Address, StandardTokenType.Custom);
+    function customTokenRegistered(address l1Address, address l2Address)
+        external
+        override
+        onlyEthPair
+    {
+        if (l2Address.isContract()) {
+            // This assumed token contract is initialized and ready to be used.
+            customToken[l1Address] = l2Address;
+            emit CustomTokenRegistered(l1Address, l2Address);
+        } else {
+            // deploy erc20 temporarily, but users can migrate to custom implementation once deployed
```

Comment:
"migrating" here means withdrawing from TMT and depositing into the custom implementation, right?

Reply:
Yes, or an already deployed token that then gets upgraded to trigger the register custom token, users can migrate into the new token.